fsw/cfe-core/src/es/cfe_es_resource.c

```diff
@@ -0,0 +1,428 @@
+    }
+    else
+    {
+        LibRecPtr = NULL;
+    }
+
+    return LibRecPtr;
+}
+
+/*********************************************************************/
+/*
+ * CFE_ES_LocateTaskRecordByID
+ *
+ * For complete API information, see prototype in header
+ */
+CFE_ES_TaskRecord_t *CFE_ES_LocateTaskRecordByID(CFE_ES_ResourceID_t TaskID)
+{
+    CFE_ES_TaskRecord_t *TaskRecPtr;
+    uint32 Idx;
+
+    if (CFE_ES_TaskID_ToIndex(TaskID, &Idx) == CFE_SUCCESS)
+    {
+        TaskRecPtr = &CFE_ES_Global.TaskTable[Idx];
+    }
+    else
+    {
+        TaskRecPtr = NULL;
+    }
+
+    return TaskRecPtr;
+}
+
+/*********************************************************************/
+/*
+ * CFE_ES_LocateCounterRecordByID
+ *
+ * For complete API information, see prototype in header
+ */
+CFE_ES_GenCounterRecord_t* CFE_ES_LocateCounterRecordByID(CFE_ES_ResourceID_t CounterID)
+{
+    CFE_ES_GenCounterRecord_t *CounterRecPtr;
+    uint32 Idx;
+
+    if (CFE_ES_CounterID_ToIndex(CounterID, &Idx) == CFE_SUCCESS)
+    {
+        CounterRecPtr = &CFE_ES_Global.CounterTable[Idx];
+    }
+    else
+    {
+        CounterRecPtr = NULL;
+    }
+
+    return CounterRecPtr;
+}
+
+/*********************************************************************/
+/*
+ * CFE_ES_GetTaskRecordByContext
+ *
+ * For complete API information, see prototype in header
+ *
+ * This function does additional validation on the task record
+ * and should only be called when global data is locked.
+ */
+CFE_ES_TaskRecord_t *CFE_ES_GetTaskRecordByContext(void)
+{
+    CFE_ES_TaskRecord_t *TaskRecPtr;
+    CFE_ES_ResourceID_t TaskID;
+
+    /*
+    ** Use the OS task ID to get the ES task record
+    */
+    TaskID = CFE_ES_ResourceID_FromOSAL(OS_TaskGetId());
+    TaskRecPtr = CFE_ES_LocateTaskRecordByID(TaskID);
+
+    /*
+     * Confirm that the entry is actually a match (this requires/assumes
+     * the global data is locked).
+     *
+     * If not a match, return NULL.
+     */
+    if (!CFE_ES_TaskRecordIsMatch(TaskRecPtr, TaskID))
+    {
+        TaskRecPtr = NULL;
+    }
+
+    return TaskRecPtr;
+}
+
+/*********************************************************************/
+/*
+ * CFE_ES_GetAppRecordByContext
+ *
+ * For complete API information, see prototype in header
+ */
+CFE_ES_AppRecord_t *CFE_ES_GetAppRecordByContext(void)
+{
+    CFE_ES_AppRecord_t *AppRecPtr;
+    CFE_ES_TaskRecord_t *TaskRecPtr;
+
+    /*
+    ** Step 1: Get the task record
+    */
+    TaskRecPtr = CFE_ES_GetTaskRecordByContext();
+    if (TaskRecPtr != NULL)
+    {
+        /*
+        ** Step 2: get the Application ID for the current task
+        */
+        AppRecPtr = CFE_ES_LocateAppRecordByID(TaskRecPtr->AppId);
+
+        /*
+         * Confirm that the entry is actually a match (this requires/assumes
+         * the global data is locked).
+         *
+         * If not a match, return NULL.
+         */
+        if (!CFE_ES_AppRecordIsMatch(AppRecPtr, TaskRecPtr->AppId))
```

Comment:
Same here. Will this ever catch a problem? Maybe a corrupted table?

Reply:
Yep - same as other spot - "Locate" just finds where it is _supposed_ to be, "IsMatch" actually checks it.